libvips/conversion/cast.c

```diff
@@ -129,7 +129,7 @@ G_DEFINE_TYPE( VipsCast, vips_cast, VIPS_TYPE_CONVERSION );
  * CAST_UINT() for an INT source, and vice versa.
  */
 #define CAST_UINT( X ) VIPS_MAX( 0, (X) )
-#define CAST_INT( X ) VIPS_MIN( (X), INT_MAX )
+#define CAST_INT( X ) VIPS_MIN( (int) (X), INT_MAX )
```

Comment:
I'm not sure about this -- `CAST_INT` is used to go from unit -> int, so casting to int before clipping the range can produce a large negative number (what the clip is supposed to avoid).

How about `((int) VIPS_MIN( (X), INT_MAX ))` instead, ie. casting after clipping the range?

Reply:
Ah, I tried that too. But unfortunately, it still produces a compiler warning with that change.

<details>
  <summary>Details</summary>

```
cast.c:400:4: warning: implicit conversion from 'int' to 'float' changes value from 2147483647 to 2147483648 [-Wimplicit-const-int-float-conversion]
                        BAND_SWITCH_INNER( float,
                        ^~~~~~~~~~~~~~~~~~~~~~~~~
cast.c:312:32: note: expanded from macro 'BAND_SWITCH_INNER'
                INT( ITYPE, signed int, int, CAST_INT ); \
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
cast.c:225:10: note: expanded from macro 'CAST_FLOAT_INT'
                q[x] = CAST( p[x] ); \
                       ^~~~~~~~~~~~
cast.c:132:45: note: expanded from macro 'CAST_INT'
#define CAST_INT( X ) ((int) VIPS_MIN( (X), INT_MAX ))
                             ~~~~~~~~~~~~~~~^~~~~~~~~
/emsdk/upstream/emscripten/system/include/libc/limits.h:30:18: note: expanded from macro 'INT_MAX'
#define INT_MAX  0x7fffffff
                 ^~~~~~~~~~
../../libvips/include/vips/util.h:55:34: note: expanded from macro 'VIPS_MIN'
#define VIPS_MIN( A, B ) ((A) < (B) ? (A) : (B))
                              ~  ^
cast.c:400:4: warning: implicit conversion from 'int' to 'float' changes value from 2147483647 to 2147483648 [-Wimplicit-const-int-float-conversion]
                        BAND_SWITCH_INNER( float,
                        ^~~~~~~~~~~~~~~~~~~~~~~~~
cast.c:312:32: note: expanded from macro 'BAND_SWITCH_INNER'
                INT( ITYPE, signed int, int, CAST_INT ); \
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
cast.c:225:10: note: expanded from macro 'CAST_FLOAT_INT'
                q[x] = CAST( p[x] ); \
                     ~ ^~~~~~~~~~~~
cast.c:132:45: note: expanded from macro 'CAST_INT'
#define CAST_INT( X ) ((int) VIPS_MIN( (X), INT_MAX ))
                                            ^~~~~~~
/emsdk/upstream/emscripten/system/include/libc/limits.h:30:18: note: expanded from macro 'INT_MAX'
#define INT_MAX  0x7fffffff
                 ^~~~~~~~~~
../../libvips/include/vips/util.h:55:46: note: expanded from macro 'VIPS_MIN'
#define VIPS_MIN( A, B ) ((A) < (B) ? (A) : (B))
                                             ^
cast.c:414:4: warning: implicit conversion from 'int' to 'float' changes value from 2147483647 to 2147483648 [-Wimplicit-const-int-float-conversion]
                        BAND_SWITCH_INNER( float,
                        ^~~~~~~~~~~~~~~~~~~~~~~~~
cast.c:312:32: note: expanded from macro 'BAND_SWITCH_INNER'
                INT( ITYPE, signed int, int, CAST_INT ); \
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
cast.c:235:10: note: expanded from macro 'CAST_COMPLEX_INT'
                q[x] = CAST( p[0] ); \
                       ^~~~~~~~~~~~
cast.c:132:45: note: expanded from macro 'CAST_INT'
#define CAST_INT( X ) ((int) VIPS_MIN( (X), INT_MAX ))
                             ~~~~~~~~~~~~~~~^~~~~~~~~
/emsdk/upstream/emscripten/system/include/libc/limits.h:30:18: note: expanded from macro 'INT_MAX'
#define INT_MAX  0x7fffffff
                 ^~~~~~~~~~
../../libvips/include/vips/util.h:55:34: note: expanded from macro 'VIPS_MIN'
#define VIPS_MIN( A, B ) ((A) < (B) ? (A) : (B))
                              ~  ^
cast.c:414:4: warning: implicit conversion from 'int' to 'float' changes value from 2147483647 to 2147483648 [-Wimplicit-const-int-float-conversion]
                        BAND_SWITCH_INNER( float,
                        ^~~~~~~~~~~~~~~~~~~~~~~~~
cast.c:312:32: note: expanded from macro 'BAND_SWITCH_INNER'
                INT( ITYPE, signed int, int, CAST_INT ); \
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
cast.c:235:10: note: expanded from macro 'CAST_COMPLEX_INT'
                q[x] = CAST( p[0] ); \
                     ~ ^~~~~~~~~~~~
cast.c:132:45: note: expanded from macro 'CAST_INT'
#define CAST_INT( X ) ((int) VIPS_MIN( (X), INT_MAX ))
                                            ^~~~~~~
/emsdk/upstream/emscripten/system/include/libc/limits.h:30:18: note: expanded from macro 'INT_MAX'
#define INT_MAX  0x7fffffff
                 ^~~~~~~~~~
../../libvips/include/vips/util.h:55:46: note: expanded from macro 'VIPS_MIN'
#define VIPS_MIN( A, B ) ((A) < (B) ? (A) : (B))
                                             ^
4 warnings generated.
```
</details>

Perhaps it's just a false-positive, I can undo that change if you want.

Reply:
Alternatively, this patch also seems to fix those warnings:
```diff
diff --git a/libvips/conversion/cast.c b/libvips/conversion/cast.c
index 1111111..2222222 100644
--- a/libvips/conversion/cast.c
+++ b/libvips/conversion/cast.c
@@ -222,7 +222,7 @@ G_DEFINE_TYPE( VipsCast, vips_cast, VIPS_TYPE_CONVERSION );
 	OTYPE * restrict q = (OTYPE *) out; \
 	\
 	for( x = 0; x < sz; x++ ) \
-		q[x] = CAST( p[x] ); \
+		q[x] = CAST( (int) p[x] ); \
 }
 
 /* Cast complex types to an int type. Just take the real part.
@@ -232,7 +232,7 @@ G_DEFINE_TYPE( VipsCast, vips_cast, VIPS_TYPE_CONVERSION );
 	OTYPE * restrict q = (OTYPE *) out; \
 	\
 	for( x = 0; x < sz; x++ ) { \
-		q[x] = CAST( p[0] ); \
+		q[x] = CAST( (int) p[0] ); \
 		p += 2; \
 	} \
 }
 ```

Reply:
Oh, that won't work.

Instead, I think we should just remove the `CAST` from `CAST_FLOAT_INT` and `CAST_COMPLEX_INT`. You don't need to clip range before casting float -> int anyway. This seems to fix the warning on clang-10 for me.

Reply:
OK, resolved with 7fecd79. Thanks!